### PR TITLE
dictionary applet: allow building in-process, fix the popup response dialog

### PR DESCRIPTION
**Pull request by lukefromdc**
*In-process is needed for wayland support
*this also fixes an unresponsive response dialog out of process
*Depends on https://github.com/mate-desktop/mate-panel/pull/1400 for in-process builds to run

### CODE REVIEW

**Comment by lukefromdc** [comment]
*Note that current git master is broken, the response dialog does not come up when the menu entry to look up selected text is selected.
*The change in screen size discovery in this PR seems to fix the response dialog both in-process (x11) and out of process as well. This was found by chance when testing both configurations. 
*In-process will run on wayland, but the entry is not responsive yet. That will require further changes to `mate_panel_applet_request_focus()` in libmate-panel-applet to create a wayland code path.

 https://github.com/mate-desktop/mate-panel/pull/1400 fixes the x11 codepath which in mate-panel from current git master will segfault if called from an in-process applet under x11. Exits if not in x11, but in-process and under  x11 this function in master will continue until the code calls for using the GtkPlug that exists only out of process to find the (now only once since GTK 3.22) screen.

**Comment by raveit65** [comment]
> *Note that current git master is broken, the response dialog does not come up when the menu entry to look up selected text is selected. 

What do you mean with response dialog?
When i select the entry in menu than it searches for the text and show me the result.
I do not use the applet and i don't know it's normal behavior.

**Comment by lukefromdc** [comment]
I normally don't use this either but I am trying to get as many features as possible to be usable in wayland sessions. The issue I had with the dialog box containing the response to the dictionary lookup turned out to be only with the panel on the bottom or right, in both cases it is rendering offscreen. I proved that by having it in the upper of two bottom panels, the top of the dialog showed up over the lower bottom panel in that test, and with a panel on the right the leftmost part of it appeared,

**Comment by raveit65** [comment]
Ok, i can reproduce the issue with bottom and right panel. I am using a vertical 2 monitor setup and the response dialog is simple on the second monitor.

**Comment by raveit65** [comment]
And i can confirm that this PR fixes `response dialog` issue when applet is build out-of-process.

**Comment by raveit65** [comment]
But when building applet in-process plus using the mate-panel PR loading the applet to the panel fails. Tested with x11.
I will try to debug it.
Bildschirmfoto zu 2023-08-11 10-31-02

**Comment by raveit65** [comment]
When running mate-panel from terminal i got this error message when adding the applet.
```
[rave@mother ~]$ mate-panel --replace
cut>

** (mate-panel:165466): WARNING **: 10:48:01.581: Failed to load applet DictionaryAppletFactory::DictionaryApplet: /usr/lib64/mate-utils/libmate-dictionary-applet.so: cannot open shared object file: No such file or directory
```
It looks for the development library of the applet but mate-utils-devel wasn't installed.
If i install the -devel package i can load the applet well.
I checked several spec files of my rpms and <lib>.so is always  the devel lib.
IE: marco:
```
%files libs -f %{name}.lang
%{_libdir}/libmarco-private.so.2*
%{_datadir}/glib-2.0/schemas/org.mate.marco.gschema.xml

%files devel
%{_bindir}/marco-window-demo
%{_includedir}/marco-1
%{_libdir}/libmarco-private.so
<cut>
```
In other applet packages like mate-applets or mate-media i do not have an extra -devel package.
So i never ran in this problematic.

**Comment by raveit65** [comment]
With mate-sensors-applet which has a -devel package which isn't installed the problem doesn't exists.
```
%files -f %{name}.lang
<cut>
%{_libdir}/libmate-sensors-applet-plugin.so.*
<cut>

%files devel
%{_libdir}/libmate-sensors-applet-plugin.so
%{_includedir}/mate-sensors-applet/
```

**Comment by lukefromdc** [comment]
`lib-name.so` is the core of the applet and will be the core of many packages. Seems to me should not be in the -devel package it should be in the main package, or else the main package should depend on the devel package as a quick workaround.  

Thus the breakage you are getting comes from in-process building a shared object, your RPMs moving this to the -devel package, and if that is not installed the applet breaks. In Debian, all of mate-utils is one package, so I am not in position to debug that one.

**Comment by raveit65** [comment]
```
[root@mother ~]# ls -la /usr/lib64/mate-utils/
insgesamt 364
drwxr-xr-x    2 root root   4096 11. Aug 19:24 .
dr-xr-xr-x. 307 root root 270336 11. Aug 19:24 ..
lrwxrwxrwx    1 root root     34 15. Mär 01:00 libmate-dictionary-applet.so -> libmate-dictionary-applet.so.0.0.0
lrwxrwxrwx    1 root root     34 15. Mär 01:00 libmate-dictionary-applet.so.0 -> libmate-dictionary-applet.so.0.0.0
-rwxr-xr-x    1 root root  90648 15. Mär 01:00 libmate-dictionary-applet.so.0.0.0
```
Of course  libmate-dictionary-applet.so is the main lib and the other files are linked with it.
But it's a convention that *.so is the devel packages. I don't think that fedora guidelines are different to debian package guide lines in this case, and debian packager will run in the same problem.
Of course i can move the -devel named lib to main package, but this is more a sticking plaster to fix the real issue.
Well, i tried to fix it fo myself but all attempt fails.
I am still wondering what is different here to all other packages.
Also, that  `-lm` is used in makefile is weird. This indicates that something is wrong here.
@mate-desktop/core-team 
Any idea or statement?

**Comment by lukefromdc** [comment]
I don't know enough about distro packaging even in Debian to work on it, I always build 1 source package -> 1 binary package, which is what Debian does with the mate-utils package.

**Comment by lukefromdc** [comment]
I also looked at the applets in mate-power-manager: easy enough to port to in-process, but mate-power-manager itself is full of Xorg-specific code so no dice there without a new backend.

**Comment by lukefromdc** [comment]
I looked at an existing Debian package I use (libavformat) and the actual .so file is in the main package under it's versioned name, a non-versioned libavformat.so symlink is added in libavformat-dev

**Comment by raveit65** [comment]
I moved  the non-versioned .so file to mate-dictionary main fedora RPM package. It's OK for me.
In the end it's a minor issue and i can't fix it for myself.

**Comment by raveit65** [comment]
Under wayfire i can't add a text to the search entry and the response dialog is a x-wayland window.
Is that the current expected behavior?

File: mate-dictionary/src/gdict-applet.c [hunk 1]
```diff
@@ -1214,7 +1214,7 @@ gdict_applet_factory (MatePanelApplet *applet,
 }
 
 /* this defines the main () for the applet */
-MATE_PANEL_APPLET_OUT_PROCESS_FACTORY ("DictionaryAppletFactory",
+PANEL_APPLET_FACTORY ("DictionaryAppletFactory",
 			     GDICT_TYPE_APPLET,
 			     "mate-dictionary-applet",
 			     gdict_applet_factory,
```
Comment (raveit65) [inline mate-dictionary/src/gdict-applet.c:1220]: Indents.
```suggestion
PANEL_APPLET_FACTORY ("DictionaryAppletFactory",
		      GDICT_TYPE_APPLET,
		      "mate-dictionary-applet",
		      gdict_applet_factory,
```

**Comment by raveit65** [comment]
> *In-process will run on wayland, but the entry is not responsive yet. That will require further changes to mate_panel_applet_request_focus() in libmate-panel-applet to create a wayland code path.

Ok, got it. PR does what it promise.

File: mate-dictionary/src/gdict-applet.c (hunk 1 above)
Comment (lukefromdc) [reply mate-dictionary/src/gdict-applet.c]: Indents should now be fixed. This is another mixed tabs and spaces file, and again I got burned on original  8 space tabs vs Pluma defaulting to 4 space tabs. I need to remember to reset that whenever editing probably any file containing tabs in MATE

**Comment by lukefromdc** [comment]
If that window is opening in Xwayland in a wayfire session, that probably indicates an x11 specific call in it. I will see if there is a simple fix and add it to this PR if I find one. We could lock it to the applet with gtk-layer-shell at the price of depending on  gtk-layer-shell for wayland builds, or leave it opening in the center of the screen.

**Comment by raveit65** [comment]
> We could lock it to the applet with gtk-layer-shell at the price of depending on gtk-layer-shell for wayland builds, or leave it opening in the center of the screen.

I am totally passionless with this applet, maybe it is OK to avoid another dependency to gtk-layer-shell for the moment. 
This can be done later if a user request it.

File: mate-dictionary/src/gdict-applet.c (hunk 1 above)
Comment (raveit65) [reply mate-dictionary/src/gdict-applet.c]: looks good.

**Comment by lukefromdc** [comment]
I just tested with Xwayland disabled, and the window still came up. Looks like it's just a wayland new window, decorated and in the center. Text is entered in it by the program if you try to look up the empty string in the entry window on the panel, so it is working. The focus issue we can leave for another PR.

File: mate-dictionary/src/gdict-applet.c (hunk 1 above)
Comment (raveit65) [reply mate-dictionary/src/gdict-applet.c]: I can replace tabs with spaces in this file when you are done with it (this and future wayland PRs).

Comment (lukefromdc) [reply mate-dictionary/src/gdict-applet.c]: Nice, that will make future work much easier

**Review by raveit65 — APPROVED** [review]
LGTM

**Comment by raveit65** [comment]
> I just tested with Xwayland disabled, and the window still came up. Looks like it's just a wayland new window, decorated and in the center. Text is entered in it by the program if you try to look up the empty string in the entry window on the panel, so it is working. The focus issue we can leave for another PR.

Good, for better understanding, why x-wayland is used when it isn't needed?

**Comment by lukefromdc** [comment]
I have no idea, there was no change in the appearance of the window on my end. Wonder if it's just being wrongly labelled? How are you determining that it is running in Xwayland?

**Comment by lukefromdc** [comment]
For ease of later debugging, I will merge this, then further fixes will be in a separate PR. This PR just makes wayland running possible and fixes an issue in running in X11. Probably best to split this from any more complex wayland work

**Comment by raveit65** [comment]
> I have no idea, there was no change in the appearance or behavior of the window on my end with xwayland enabled or disabled. Wonder if it's just being wrongly labelled? How are you determining that it is running in Xwayland?

It looks like a xwayland window, or not....?
dictionary_applet

**Comment by raveit65** [comment]
I will disable xwayland in wayfire.ini to ssee how it works.

**Comment by lukefromdc** [comment]
I've had lots of wayland windows look exactly like that, notably ones that needed to be locked down with gtk-layer-shell

**Comment by raveit65** [comment]
Disable xwayland doesn't make any different for the window. Maybe this decoration is used for undecorated windows?
But when i disable xwayland in wayfire.ini and try to open this window again from the panel without restarting the session, the panel crashes.

**Comment by lukefromdc** [comment]
Than we have a difference in behavior we need to find. One possible way might be to remove all x11-related includes from the file for that window and see what breaks. Whatever it is, the string "x11" didn't show up in the source, so it must be some Xorg related function that is "x<something>"

**Comment by raveit65** [comment]
It is simmply the decoration for a none-decorated window.
This brings the decoration back.
```
From ff4f436d50033424a1f9ad8850ca1a5d2c8a1dc5 Mon Sep 17 00:00:00 2001
From: raveit65 <mate@raveit.de>
Date: Sat, 12 Aug 2023 22:14:53 +0200
Subject: [PATCH] decorate applet window

---
 mate-dictionary/src/gdict-aligned-window.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/mate-dictionary/src/gdict-aligned-window.c b/mate-dictionary/src/gdict-aligned-window.c
index 42a00cc9..6f17ddbf 100644
--- a/mate-dictionary/src/gdict-aligned-window.c
+++ b/mate-dictionary/src/gdict-aligned-window.c
@@ -102,8 +102,8 @@ gdict_aligned_window_init (GdictAlignedWindow *aligned_window)
 #if 0
   gtk_window_set_modal (window, TRUE);
 #endif
-  gtk_window_set_decorated (window, FALSE);
-  gtk_window_set_type_hint (window, GDK_WINDOW_TYPE_HINT_DOCK);
+  gtk_window_set_decorated (window, TRUE);
+  gtk_window_set_type_hint (window, GDK_WINDOW_TYPE_HINT_DIALOG);
 }
 
 static void
-- 
2.41.0
```
Not sure if i choose the best GDK_WINDOW_TYPE_HINT and this needs behind a wayland clause.
But it works.

**Comment by lukefromdc** [comment]
I'll set that up wayland-only as another PR, it's simple and a problem solver

**Comment by raveit65** [comment]
It look like this
dictionary_applet_better

**Comment by lukefromdc** [comment]
Your patch works, but with or without it I found another issue: if the user closes the window with the close button, the applet can't rebuilt it and becomes nonresponsive. Will test a fix using gtk-layer-shell, as always that would have to be a build-time option

**Comment by raveit65** [comment]
Maybe we choose another GDK_WINDOW_TYPE_HINT without a close button.
Or try to disable 
```
 #if 0
   gtk_window_set_modal (window, TRUE);
 #endif
```
I think `modal` means no window buttons.

**Comment by lukefromdc** [comment]
Modal attaches a window to it's parent and might work

**Comment by raveit65** [comment]
https://docs.gtk.org/gtk3/method.Window.set_deletable.html
I will try to disable the the close button with `gtk_window_set_deletable ()`, which works well with gtk-inspector.

**Comment by lukefromdc** [comment]
Modal didn't change anything in location or button layout in the dialog, and still worse I get a segfault taking down the whole panel if I clicked on the icon after closing the window. 

Turns out in wayland GDK_WINDOW_TYPE_HINTS are apparently ignored
https://github.com/swaywm/sway/issues/4655

**Comment by lukefromdc** [comment]
Same results with    
`gtk_window_set_deletable (window, FALSE);`

In early testing using gtk layer shell works, just haven't finished aligning the window and need to make that method build time selectable

**Comment by raveit65** [comment]
```
diff --git a/mate-dictionary/src/gdict-aligned-window.c b/mate-dictionary/src/gdict-aligned-window.c
index 42a00cc9..5350dd2c 100644
--- a/mate-dictionary/src/gdict-aligned-window.c
+++ b/mate-dictionary/src/gdict-aligned-window.c
@@ -102,8 +102,9 @@ gdict_aligned_window_init (GdictAlignedWindow *aligned_window)
 #if 0
   gtk_window_set_modal (window, TRUE);
 #endif
-  gtk_window_set_decorated (window, FALSE);
+  gtk_window_set_decorated (window, TRUE);
   gtk_window_set_type_hint (window, GDK_WINDOW_TYPE_HINT_DOCK);
+  gtk_window_set_deletable (window, FALSE);
 }
 
 static void
-- 
2.41.0
This works for me, Close button is disable and opening/closing the dock with the panel icon works without crashing the panel.
One minor issue with this is that the window becomes huger with every opening.
But i noticed the same weid behaviour with mate-terminal too when minimal/restore frm tasklist.

Any way up to you to decide what to do.

**Comment by lukefromdc** [comment]
Nearly done with the layer-shell work, and a retest of your patch didn't work here. Worked for you, something is different but I think I can get better results this way and hopefully also get the entry to work

**Comment by lukefromdc** [comment]
https://github.com/mate-desktop/mate-utils/tree/wayland-dictionary
will get a PR submitted as soon as I get the entry to work properly

**Comment by lukefromdc** [comment]
I was able to prove with a g_message() statement that when we cannot use mate_panel_applet_request_focus() as is the case in wayland, clicking on the entry is recognized, but it cannot grab the keyboard or accept a paste. Anything valid put into gdict-applet.c function `gdict_applet_entry_button_press_event_cb` will run but I have not been able to get the entry to accept input, making the applet useless.

The freestanding mate-dictionary applet works just fine in wayland

**Comment by raveit65** [comment]
I tested it and position of the window is fine.
The GtkEntry can be fixed in future, ihmo, otherwise all the in-process work  is useless.
When i try to type something in the GtkEntry always the mini-search window of caja opens.
It seems the GtkEntry doesn't have focus.
The mini-search should be disable for caja-desktop.

**Comment by lukefromdc** [comment]
I will keep looking for that, and will open a PR for the existing wayland work. That entry is the only problem left, if we don't find a fix before 1.28 is released we could mark it as "experimental" in the help text in configure.ac and in an inline comment at the entry;

**Comment by lukefromdc** [comment]
https://github.com/mate-desktop/mate-utils/pull/354

It seems that the focus either "falls through" the applet to the panel or is intercepted by the panel, even though if I add `gtk_widget_unrealize` to the callback the entry will disappear when clicked as one would expect it to. 

Note that mate-panel's `panel_applet_request_focus` has no other user within MATE's own packages, and uses low level Xorg code to get the focus for this entry. My guess is it was written when for some reason an entry in the panel could not be focussed when it was written.

Worst case we could change the codepath in wayland to replace the entry with another icon, one which would popup a box to put the text into that can be focussed. That would be wayland-only of course

**Pull request merged**